File: ruby/lib/tree.rb
# frozen_string_literal: true

require_relative './node'
require 'pry'

class Tree
  attr_reader :root, :size

  NODE_CLASS = Node

  def preorder_walk
    return unless root
    root.pre_order_traverse(&Proc.new)
  end

  def inorder_walk
    return unless root
    root.in_order_traverse(&Proc.new)
  end

  def postorder_walk
    return unless root
    root.post_order_traverse(&Proc.new)
  end

  def initialize node = nil
    # TODO: find out why both versions pass all
    # the specs. That's bad. One should pass,
    # the other fail.
    # @root = node ? node : NODE_CLASS
    @root = node ? node : NODE_CLASS.new
    @size = 1
  end

  def empty?
    root.nil? ? true : false
  end

  def insert node
    @root.insert node
    @size += 1
  end

  def invert
    return if root.nil?
    preorder_walk do |node|
      temp = node.right
      node.right = node.left
      node.left = temp
    end
  end

  # From CLRS (3rd Edition) p. 296, "transplant" a node's
  # location in the tree. The idea is to replace u with v.
  # This function would be better named `reset_parent` because
  # that's all it does. Resetting child links is done in the
  # delete method.
  def transplant u, v
    # Consider "ruby-izing the following. Rubocop doesn't like it.
    if u.parent.nil? # u is a root node
      @root = v
    elsif u == u.parent.left # u is a left child of its parent
      u.parent.left = v
    else # u is a right child of its parent
      u.parent.right = v
    end

    # if !v.nil?
    v&.parent = u.parent
    # v.parent = u.parent unless v.nil?
    # end
    v
  end

  def transplant node, child
    if node.parent.nil?
      @root = child
    elsif node.left_child?
      node.parent.left = child
    else # right child
      node.parent.right = child
    end

    child&.parent = node.parent
    # return value is given in book, but not used
    # child
  end

  def delete_clrs3 key
    z = search key
    if z.left.nil?
      transplant z, z.right
    elsif z.right.nil?
      transplant z, z.left
    else
      y = z.right.minimum
      if y.parent != z
        transplant y, y.right
        y.right = z.right
        y.right.parent = y
      end
      transplant z, y
      y.left = z.left
      y.left.parent = y
    end
    @size -= 1
    z.left = z.right = z.parent = nil
    z
  end

  # This is kind of nasty. It turns out that rooting a tree
  # in some sort of container is both a crutch and a constraint.
  # The big hassle is the edge case induced by deleting the root
  # node.
  def delete key
    left = @root.left
    right = @root.right
    deleted_node = @root.delete key
    @size -= 1
    # This is arbitrary, we just pick a side.
    @root = left || right if deleted_node == @root
    deleted_node
  end

  def delete_by_key key
    delete_clrs3 key
  end

  def list_keys
    root&.list_keys || []
  end

  def collect _collector
    root&.collect [] || []
  end

  def preorder_keys
    collector = []
    preorder_walk { |node| collector << node.key }
    collector
  end

  def search key
    @root.find key
  end

  def height
    root.nil? ? 0 : root.height
  end

  def full?
    root.full?
  end

  def bst?
    root.bst?
  end

  def successor node
    root.successor node
  end

  def predecessor node
    root.predecessor node
  end

  def maximum
    @root.maximum
  end

  def minimum
    @root.minimum
  end

  def to_hash
    root.to_hash
  end

  def to_json
    root.to_json
  end

  def to_json_file filename
    File.open(filename, 'w') do |file|
      file.write root.to_json
    end
  end

  def self.from_json_file filename
    require 'json'
    file = File.read(filename)
    hash = JSON.parse(file)
    from_hash(hash)
  end

  def self.from_hash hash
    Tree.new(NODE_CLASS.build_from_hash(hash))
  end

  def get_next_row current_row
    next_row = []
    current_row.each do |node|
      next_row << node.left
      next_row << node.right
    end
    next_row.compact
  end

  # TODO: rename to breadth_first_traverse
  # TODO: consider add code to test if root.nil?
  # TODO: rewrite this using a single method with a stack.
  def bfsearch
    rows = [[root]]
    rows << get_next_row(rows.last) until rows.last.empty?
    rows.flatten.map(&:key)
  end
end

File: python/lib/tree.py
import json
from node import *

class Tree(object):

    def __init__(self, node):
        self.root = node

    def is_empty(self):
        return bool(self.root is None)

    def insert(self, node):
        self.root.insert(node)

    def find(self, value):
        return self.root.find(value)

    def is_present(self, value):
        return self.root.is_present(value)

    def list_keys(self):
        if self.root is None:
            return []
        else:
            return self.root.list_keys()

    def collect(self, collector):
        self.root.collect(collector)

    def size(self):
        if self.root is None:
            return 0
        return self.root.size()

    def maximum(self):
        return self.root.maximum()

    def minimum(self):
        return self.root.minimum()

    def is_bst(self):
        return self.root.is_bst()

    def height(self):
        if self.root is None:
            return 0
        return self.root.height()

    def successor(self, node):
        return self.root.successor(node)

    def predecessor(self, node):
        return self.root.predecessor(node)

    # A better name for `transplant` might be `reset_parent`
    def transplant(self, u, v):
        if u.parent is None:
            self.root = v
        elif u == u.parent.left:
            u.parent.left = v
        else:
            u.parent.right = v

        if v is not None:
            v.parent = u.parent

    def clrs_delete(self, key):
        z = self.find(key)

        if z.left is None:
            self.transplant(z, z.right)
        elif z.right is None:
            self.transplant(z, z.left)
        else:
            y = z.right.minimum()
            if y.parent != z:
                self.transplant(y, y.right)
                y.right = z.right
                y.right.parent = y

            self.transplant(z, y)
            y.left = z.left
            y.left.parent = y


    # Refactored and renamed from CLR. The CLR code is a little
    # difficult to follow as it has a lot of conditional statements
    # and single letter temporary nodes (x, y, z).
    def clr_delete(self, node):
        # for sanity checking as delete method is refactored
        # return self.clr_delete_original(node)

        # z = self.root.find(node.key)

        z = node

        if z.left and z.right:
            node_to_delete = self.root.successor(z)
            # This is where to swap the nodes
            z.key = node_to_delete.key
        else:
            node_to_delete = z

        if node_to_delete.left:
            x = node_to_delete.left
        else:
            x = node_to_delete.right

        if x:
            # TODO: This line works with x.p, needs to be better tested
            # Try deleting x or x.p sucessively to figure out exactly
            # how it works.
            # x.p = node_to_delete.parent
            x.parent = node_to_delete.parent

        if node_to_delete.parent:
            if node_to_delete == node_to_delete.parent.left:
                node_to_delete.parent.left = x
            else:
                node_to_delete.parent.right = x
        else:
            self.root = x

        return node_to_delete


    # algorithm originally taken from CLR p. 253
    # parent pointers are part of CLR's definition of
    # node in BST. The names have been mostly kept the
    # same, however, 'parent' is used instead of 'p' for
    # better readability in refactored delete implementation.
    def clr_delete_original(T, z):
        if z.left is None or z.right is None:
            y = z
        else:
            y = T.root.successor(z)

        if y.left:
            x = y.left
        else:
            x = y.right

        if x is not None:
            x.parent = y.parent

        if y.parent is None:
            T.root = x
        else:
            if y == y.parent.left:
                y.parent.left = x
            else:
                y.parent.right = x

        if y != z:
            z.key = y.key

        return y

    def to_json(self):
        return True
        # print json.dumps(['foo', {'bar': ('baz', None, 1.0, 2)}])
